contracts/random/PRNG.sol

```diff
@@ -38,31 +41,32 @@ library PRNG {
     }
 
     /**
-    @dev Hashes seed||counter, placing it in the entropy word, and resets the
-    remaining bits to 256. Increments the counter BEFORE the refill (ie 0 is
-    never used) as this simplifies round-tripping with store() and loadSource()
-    because the stored counter state is the same as the one used for deriving
-    the entropy pool.
+    @dev Computes the next PRN in entropy using a lag-1 multiply-with-carry
+    algorithm and resets the remaining bits to 128.
+    `nextNumber = (factor * number + carry) mod 2**128`
+    `nextCarry  = (factor * number + carry) //  2**128`
      */
     function _refill(Source src) private pure {
         assembly {
-            let ctr := add(src, COUNTER)
-            mstore(ctr, add(1, mload(ctr)))
-            mstore(add(src, ENTROPY), keccak256(src, 0x40))
-            mstore(add(src, REMAIN), 256)
+            let carryAndNumber := mload(src)
+            let rand := and(carryAndNumber, 0xffffffffffffffffffffffffffffffff)
+            let carry := shr(128, carryAndNumber)
+            let tmp := add(mul(MWC_FACTOR, rand), carry)
```

Comment:
Interesting! So the carry is implicitly stored in the upper half of the word as a consequence of the algorithm?

Reply:
Yes. That's where part of the efficiency comes from because both the number and the carry can be computed/store in one computation/word. The downside is that the random number is only 128 bit long.